release_util/management/commands/__init__.py

```diff
@@ -132,164 +134,215 @@ def state(self):
                 if line_match.group('success') == 'OK':
-                    # A migration succeeded.
-                    self.migration_state['success'].append({
-                        'migration': migration,
-                        'output': line,
-                    })
+                    # The migration succeeded
+                    succeeded.append(migration)
+                else:
+                    # The migration failed
+                    failed = migration
+                    break
+        return succeeded, failed
 
-    def apply_all(self):
+    def _apply(self, migration=None, run_all=False):
+        """ 
+        If a migration is supplied, runs that migration and appends to state.
+        If run_all==True, runs all migrations.
+        Raises a ValueError if neither "migration" nor "run_all" are provided.
+        """
+        out = StringIO()
+        failed = traceback = None
+        
+        migrate_kwargs = {
+            'interactive': False,
+            'stdout': out,
+            'database': self._database_name,
+        }
+        if migration is not None:
+            migrate_kwargs.update({
+                'app_label': app,
+                'migration_name': migration,
+            })
+        elif not run_all:
+            raise ValueError('Either a migration must be provided or "run_all" must be True')
+
+        start = self.timer()        
+        try:
+            call_command("migrate", **migrate_kwargs)
+        except (CommandError, DatabaseError):
+            traceback = repr(traceback.format_exception(*sys.exc_info()))
+
+        end = self.timer() - start
+        successes, failure = self._parse_migrate_output(out.getvalue())
+        
+        self.migration_state.append({
+            'migration': (app, migration),
+            'duration': end - start,
+            'output': out.getvalue(),
+            'success': successes,        # [(app, migration), ...]
+            'unapplied': self._to_apply, # [(app, migration), ...]
+            'failure': failure,          # (app, migration)
+            'traceback': traceback,
+        })
+
+    def apply(self):
         """
-        Apply all the migrations together.
+        Applies all migrations that have been added.
+        Note that some migrations depend on others, so you might end up
+        running more than one.
         """
+        if self._closed:
+            raise MigrationSessionError("Can't apply applied session")
+        try:
+            while self._to_apply:
+                self._apply(migration=self._to_apply.pop(0))
+        except:
+            raise
+        finally:
+            self._closed = True
 
-        self._apply_all()
-        self._add_rollback_commands()
+    def apply_all(self):
+        """
+        Applies all Django model migrations at once, recording the result.
+        """
+        if self._closed:
+            raise MigrationSessionError("Can't apply applied session")
+        if self._to_apply:
```

Comment:
After clarification about dunders: can you clarify? I'm not sure I understand why checking states implies this method should be extra-private.